pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -444,6 +446,21 @@ func (c *Controller) InstancesByPort(hostname model.Hostname, reqSvcPort int,
 						uid = fmt.Sprintf("kubernetes://%s.%s", pod.Name, pod.Namespace)
 					}
 
+					// find in the MeshNetworks configuration what's the network
+					var network string
+					if c.Env.MeshNetworks != nil {
+						for n, v := range c.Env.MeshNetworks.Networks {
+							for _, ep := range v.Endpoints {
+								if ep.GetFromRegistry() == c.ClusterID {
+									network = n
+								}
+								if util.CidrContainsIP(ep.GetFromCidr(), ea.IP) {
```

Comment:
This seems reversed and does a ton of recomputation (and is in the hot path as well IIRC, @costinm can confirm). First, the data in the meshnetworks config doesn't change for the life of the environment, we should spend time computing the set of CIDRs once, rather than redoing it all the time. Second, the lookup should be reversed: given an IP find the CIDR it belongs to, rather than trying to match CIDRs to the address.

There are libraries in golang that implement efficient IP to CIDR matching, e.g. [cidranger](https://github.com/yl2chen/cidranger) or [patricia trees](https://github.com/kentik/patricia).

Reply:
You are right. Basically the Mesh Networks configmap can change but once I add a watcher for that configmap I can also purge the CIDR mapping here and reinit it.

The `cidranger` is nice, elegant and efficient. I changed the code to be based on it rather than my util function.